dali/pipeline/util/bounding_box_utils.h

```diff
@@ -0,0 +1,243 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.c++ copyswa
+
+#ifndef DALI_PIPELINE_UTIL_BOUNDING_BOX_UTILS_H_
+#define DALI_PIPELINE_UTIL_BOUNDING_BOX_UTILS_H_
+
+#include <limits>
+#include "dali/core/geom/box.h"
+#include "dali/core/tensor_layout.h"
+#include "dali/core/math_util.h"
+
+namespace dali {
+
+template <int ndim>
+void AxisFlip(Box<ndim, float>& box, int axis) {
+  assert(axis >= 0 && axis < ndim);
+  box.lo[axis] = 1.0f - box.hi[axis];
+  box.hi[axis] = 1.0f - box.lo[axis];
+}
+
+template <int ndim>
+void HorizontalFlip(Box<ndim, float>& box) {
+  static_assert(ndim >= 1);
+  AxisFlip(box, 0);
+}
+
+template <int ndim>
+void VerticalFlip(Box<ndim, float>& box) {
+  static_assert(ndim >= 2);
+  AxisFlip(box, 1);
+}
+
+template <int ndim>
+void DepthWiseFlip(Box<ndim, float>& box) {
+  static_assert(ndim >= 3);
+  AxisFlip(box, 2);
+}
+
+template <int ndim>
+constexpr TensorLayout DefaultLayout() {
+  assert(ndim == 2 || ndim == 3);
+  if (ndim == 3)
+    return TensorLayout{"xyzXYZ"};
+  else
+    return TensorLayout{"xyXY"};
+}
+
+template <int ndim>
+constexpr TensorLayout DefaultAnchorAndShapeLayout() {
+  assert(ndim == 2 || ndim == 3);
+  if (ndim == 3)
+    return TensorLayout{"xyzWHD"};
+  else
+    return TensorLayout{"xyWH"};
+}
+
+template <int ndim>
+float AspectRatio(const Box<ndim, float>& box, int dim0, int dim1) {
+  assert(dim0 >= 0 && dim0 < ndim);
+  assert(dim1 >= 0 && dim1 < ndim);
+  auto extent = box.extent();
+  auto extent0 = extent[dim0];
+  auto extent1 = extent[dim1];
+  return extent0 / extent1;
+}
+
+template <int ndim>
+float MinAspectRatio(const Box<ndim, float>& box) {
+  float min_ar = std::numeric_limits<float>::max();
+  for (int i = 0; i < ndim; i++) {
+    for (int j = i + 1; j < ndim; j++) {
+      min_ar = std::min(min_ar, AspectRatio(box, i, j));
+    }
+  }
+  return min_ar;
+}
+
+template <int ndim>
+float MaxAspectRatio(const Box<ndim, float>& box) {
+  float max_ar = std::numeric_limits<float>::min();
+  for (int i = 0; i < ndim; i++) {
+    for (int j = i + 1; j < ndim; j++) {
+      max_ar = std::max(max_ar, AspectRatio(box, i, j));
+    }
+  }
+  return max_ar;
+}
```

Comment:
The `AspectRatio` doesn't take any assumptions about the order of `dim0` and `dim1`, but the nested loops in Min/MaxAspectRatio indicate dim0 < dim1. Is this intended behaviour?

Reply:
the current operator has an aspect ratio range and that means the aspect ratio of x/y. To make this operator accept volumetric inputs I decided to extend this logic so that the range applies for x/y, x/z and y/z. We could enhance the way to limit AR to different ranges for different dimensions but it seems like overdoing it at this moment as there is no real use case so I decided to keep it simple

Reply:
Can you mention in a comment that those assume the aspect ratio is taken wrt to order of axes, shape[i] / shape[j] for i < j or something like that?

Reply:
This function is not used anywhere - as such, I don't think it's very useful.

Reply:
Shouldn't you skip case `i == j`?